symfony/src/Security/GoogleGroupsAuthenticator.php

```diff
@@ -69,18 +76,29 @@ public function checkCredentials($credentials, UserInterface $user)
             return true; // Bypass groups auth
         }
         $valid2fa = !($this->params->has('enforce2fa') && $this->params->get('enforce2fa')) || $user->hasTwoFactorAuth();
+        $this->session->set('authEmail', $user->getUsername());
+        if (!$valid2fa) {
+            $this->session->set('authFailureReason', '2fa');
+        } elseif (empty($user->getGroups())) {
+            $this->session->set('authFailureReason', 'groups');
+        }
```

Comment:
When we discussed earlier, I was thinking that you were going to create a new route to redirect to, which is why I suggested storing the reason in the session. Since you are rendering the template and returning the response directly in the failures method, these could just be stored in properties in this class, removing the need to write and read from the session.